conformance/tests/specialtypes_any.py

```diff
@@ -87,4 +88,99 @@ def method1(self) -> int:
 assert_type(a.method2(), Any)
 assert_type(ClassA.method3(), Any)
 
+# > When ``Any`` is present in the bases of a type,
+# > it should be considered only after all other known types in the MRO.
 
+class ClassKnown:
+
+    classvar1 = ""
+
+    def __iter__(self) -> Iterator[str]:
+        yield from self.attr1
+
+    def __init__(self):
+        self.attr1: str = ""
+
+    def method1(self) -> str:
+        return ""
+
+class AnyFirst(Any, ClassKnown):
+
+    def method2(self) -> str:
+        return ""
+
+class AnyLast(ClassKnown, Any):
+    def method2(self) -> str:
+        return ""
+
+class GetattrKnown(ClassKnown):
+    def __getattr__(self, name: str) -> int:
+        return 1
+
+class AnyFirstGetAttr(Any, GetattrKnown):
+    def method2(self) -> str:
+        return ""
+
+class AnyLastGetAttr(GetattrKnown, Any):
+    def method2(self) -> str:
+        return ""
+
+class AnySub(Any):
+    ...
+
+# primarily included to demonstrate intent that this is for the full MRO
+class AnySubFirst(AnySub, ClassKnown):
+    def method2(self):
+        ...
+
+
+
+af = AnyFirst()
+assert_type(af.method1(), str)
+assert_type(af.method2(), str)
+assert_type(af.attr1, str)
+assert_type(af.non_exist_method(), Any)
+assert_type(af.non_exist_attr, Any)
+assert_type(af.classvar1, str)
+assert_type(AnyFirst.classvar1, str)
+assert_type(iter(af), Iterator[str])
+
+al = AnyLast()
+assert_type(al.method1(), str)
+assert_type(al.method2(), str)
+assert_type(al.attr1, str)
+assert_type(al.non_exist_method(), Any)
+assert_type(al.non_exist_attr, Any)
+assert_type(al.classvar1, str)
+assert_type(AnyLast.classvar1, str)
+assert_type(iter(al), Iterator[str])
+
+af_getattr = AnyFirstGetAttr()
+
+assert_type(af_getattr.method1(), str)
+assert_type(af_getattr.method2(), str)
+assert_type(af_getattr.attr1, str)
+assert_type(af_getattr.triggers_getattr, int)
```

Comment:
This is correct, one of the bases provided provides a `__getattr__` definition

Reply:
But `__getattr__` isn't applicable here. This method is invoked only if the attribute is not present. But in this case, it is "present" because all attributes are present on an `Any` base class.

Reply:
At runtime, this is definitely invoked, runtime doesn't see Any and do something special for `__getattr__`, I've modeled this on accurately typing runtime for a known implementation of `__getattr__`

Reply:
I don't think "modeling this on runtime" with the actual runtime `Any` type is meaningful here. The point of having `Any` in the MRO is that for typing purposes it might represent any concrete type, which might have any actual attribute. If `Any` in this case represented a concrete type that provides the `triggers_getattr` attribute, then `__getattr__` would not be invoked. So for the same reason that `non_existent_attr` works and produces `Any` in these examples, `triggers_getattr` should work and also produce `Any`, because checking `Any` for the `triggers_getattr` attribute occurs before consulting `__getattr__`, and `Any` must be considered to provide all attributes.

Reply:
It's worth adding that the original reason for Any being allowed to be subclassed at runtime involved a getattr definition. If we *don't* do this, then the order matters in exactly the case of `__getattr__` rather than being consistent, this was added for `unittest.Mock` and similar.

Reply:
> At runtime, this is definitely invoked

You don't know whether it will be invoked at runtime. The missing class represented by `Any` could have an attribute named `triggers_getattr`. If it does, then `__getattr__` will not be invoked at runtime. By the rules of `Any`, a static type checker must assume that the class will have an attribute by this name, and it's type is `Any`.

Reply:
Okay, let me clarify: for every other magic method and attribute, we have determined that Any is not being considered until the end, under one of several rationales that all were similar but not picked, yet different constructions about compatibility.

if there is a `__getattr__` from the unknown class, it must be compatible with the one from the known class (Any's guarantee of compatibility) so this should be fine.

Reply:
> if there is a __getattr__ from the unknown class, it must be compatible with the one from the known class (Any's guarantee of compatibility) so this should be fine.

That logic doesn't apply here, because the unknown class might not have a `__getattr__` at all, but it might have a `triggers_getattr`, and we have no idea what type it might be.

Reply:
What about going the other way on this? You could specify that a class that inherits from Any must only have a `__getattr__` definition if the definition is typed to return Any

Reply:
Can someone outline or link to the motivating case with mocks that would be broken if the spec just gave `Any` on these getattr cases?

Reply:
It could be resolved that way, taking this over to discourse.

Reply:
This has nothing to do with whether the unknown class implements `__getattr__`. It has to do with whether the unknown class has an attribute called `triggers_getattr`. We must assume it does, in which case the presence of a `__getattr__` method is irrelevant.

Consider the following:

```python
class A:
    foo = "foo"

class B:
    def __getattr__(self, name: str) -> int:
        print("__getattr__")
        return 1

class C(A, B):
    pass

c = C()
reveal_type(c.foo) # str
print(c.foo) # "foo"
```

Note that `B.__getattr__` is never invoked at runtime in this case, and the type checker accordingly never simulates a call to `__getattr__` when evaluating the type of expression `c.foo`.

Now, we replace `A` with `Any`. Here again, the type checker should assume that `__getattr__` will not be invoked.

```python
class C(Any, B):
    pass

c = C()
reveal_type(c.foo) # Any
```

Reply:
(github isn't loading reviews in realtime right now for anyone viewing this after the fact)